components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -93,6 +94,37 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes here.
+    protected readonly contentCache = new Map<string, string>();
+    protected readonly updatedSinceLastWrite: string[] = [];
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.push(resource.uri);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+        return this.getServer().getUserStorageResource(req);
+    }
+
+    protected startPeriodicCacheWriter(interval: number) {
+        setInterval(async () => {
```

Comment:
There is no guarantee that this code won't hang each 10 seconds if the connection to the server is lost and then ran multiple cycle in parallel.

You can just write async code which never finishes, i.e.
```js
while(true) {
   try {
      await new Promise(resolve => setTimeout(resolve, interval * 1000));
      // do the thing
   } catch (e) {
      console.error(e)
   }
}

Reply:
The downside though that after reconnecting it will wait another 10 seconds.

Reply:
Yes. If we want to back it bulletproof this needs to switch to a chained `setTimeout` with calculating diff (since last exec). That would resolve the issue @svenefftinge pointed out (a _lot_ of calls to the same resource), plus the "one slow update" plus the "a lot of updates take longer than interval".